warden_daemon/src/storage/app_disk_manager.rs

```diff
@@ -0,0 +1,419 @@
+            .clone())
+    }
+
+    fn read_partitions(&self) -> Result<HashMap<String, Partition>, ApplicationDiskManagerError> {
+        let gpt = GptConfig::default()
+            .writable(false)
+            .initialized(true)
+            .open(&self.file_path)
+            .map_err(|err| ApplicationDiskManagerError::GPTRead(err.to_string()))?;
+
+        Ok(gpt
+            .partitions()
+            .iter()
+            .map(|(_id, value)| (value.name.clone(), value.clone()))
+            .collect())
+    }
+
+    async fn sync_file(file: std::fs::File) -> Result<(), ApplicationDiskManagerError> {
+        let mut file = File::from_std(file);
+        file.sync_all()
+            .await
+            .map_err(|err| ApplicationDiskManagerError::PartitionCreation(err.to_string()))?;
+        file.shutdown()
+            .await
+            .map_err(|err| ApplicationDiskManagerError::FileShutdown(err.to_string()))
+    }
+
+    async fn create_disk_device(&self) -> Result<std::fs::File, io::Error> {
+        let mut file = File::options()
+            .read(true)
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(&self.file_path)
+            .await?;
+        file.seek(SeekFrom::Start(self.calculate_total_size() - 1))
+            .await?;
+        file.write_all(&[0u8]).await?;
+        file.seek(SeekFrom::Start(0u64)).await?;
+        Ok(file.into_std().await)
+    }
+
+    fn calculate_total_size(&self) -> u64 {
+        self.data_part_bytes_size
+            + self.image_part_bytes_size
+            + (34 * 2 * u64::from(Self::LBA_SIZE))
+        // 68 LBA sectors for GPT
+    }
+
+    fn write_mbr(&self, mut file: impl DiskDevice) -> Result<(), ApplicationDiskManagerError> {
+        let mbr = ProtectiveMBR::with_lb_size(
+            (self.calculate_total_size() / u64::from(Self::LBA_SIZE) - 1) as u32,
+        );
+        mbr.overwrite_lba0(&mut file)
+            .map_err(|err| ApplicationDiskManagerError::PartitionMetadataWrite(err.to_string()))?;
+        Ok(())
+    }
+
+    fn create_partition(
+        gpt: &mut GptDisk,
+        name: impl AsRef<str>,
+        partition_size_bytes: u64,
+    ) -> Result<(), ApplicationDiskManagerError> {
+        gpt.add_partition(
+            name.as_ref(),
+            partition_size_bytes,
+            gpt::partition_types::LINUX_FS,
+            0,
+            None,
+        )
+        .map(|_| ())
+        .map_err(|err| ApplicationDiskManagerError::PartitionCreation(err.to_string()))
+    }
+
+    fn create_gpt_and_partitions(
+        &self,
+        mut file: impl DiskDevice,
+    ) -> Result<(), ApplicationDiskManagerError> {
+        self.write_mbr(&mut file)?;
+
+        let mut gpt = GptConfig::default()
+            .initialized(false)
+            .writable(true)
+            .logical_block_size(Self::LBA_SIZE)
+            .create_from_device(Box::new(&mut file), None)
```

Comment:
Why not create the Box somewhere higher in the call stack and make this function accept Box<dyn DiskDevice>? So that write_mbr would also accept the same pointer. What do you think?

Reply:
Done